keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,918 @@
+
+### Troubleshooting
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+
+* **What are other known failure modes?**
+  For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Using signed integers for the offset
+
+Allow for the offset to be specified as either +/-, i.e. `1.5s` could be done as either `1s + 500ms` or `2s - 500ms`
+
+Pros:
+Allows for more granular control of legacy behavior (no rounding)
+
+Cons:
+Potentially ambiguous round-tripping in a future v2 API
+### Combine old and new fields
+
+Allow for 0 second Probe, with setting millisecond. Similar to what is done with `tv_sec` and `tv_nsec` from POSIX time structs.
+
+Cons:
+Loosens validation.
```

Comment:
So this proposal would just add a new field for each existing *Seconds field, like `PeriodSeconds` and `PeriodMilliseconds` and then the actual time is the sum? This seems like a *way* more intuitive and understandable API to me. It's also simpler to enter - just one new field instead of a new struct. And follows the existing *Seconds pattern. I am not seeing the downside? Validation can be done on the total duration as well as the individual fields, can't it?

Reply:
Clients that don't recognise the new field would be liable to trigger much too early, right? How do we avoid that?

Also, how do we avoid setting a precedent that this riskier approach is the one to take.

Reply:
Yeah, that's true. The lack of versioning at this granularity is a problem. Of course the "offset" has a similar problem (inaccurate interval from old clients). Here's another (maybe bad) idea:

on write:
  - if *Milliseconds is set, then set *Seconds = ceil(*Milliseconds/1000)
  - otherwise, set *Milliseconds = *Seconds * 1000

Then, on read, both will be set, and old clients will have a slightly longer interval.